src/aiy/_drivers/_recorder.py

```diff
@@ -0,0 +1,143 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A recorder driver capable of recording voice samples from the VoiceHat microphones."""
+
+import logging
+import os
+import subprocess
+import threading
+import wave
+
+import aiy._drivers._alsa
+
+logger = logging.getLogger('recorder')
+
+
+class Recorder(threading.Thread):
+
+    """A driver to record audio from the VoiceHat microphones.
+
+    Stream audio from microphone in a background thread and run processing
+    callbacks. It reads audio in a configurable format from the microphone,
+    then converts it to a known format before passing it to the processors.
+
+    This driver accumulates input (audio samples) in a local buffer. Once the
+    buffer contains more than CHUNK_S seconds, it passes the chunk to all
+    processors. An audio processor defines a 'add_data' method that receives
+    the chunk of audio samples to process.
+    """
+
+    CHUNK_S = 0.1
+
+    def __init__(self, input_device='default',
+                 channels=1, bytes_per_sample=2, sample_rate_hz=16000):
+        """Create a Recorder with the given audio format.
+
+        The Recorder will not start until start() is called. start() is called
+        automatically if the Recorder is used in a `with`-statement.
+
+        - input_device: name of ALSA device (for a list, run `arecord -L`)
+        - channels: number of channels in audio read from the mic
+        - bytes_per_sample: sample width in bytes (eg 2 for 16-bit audio)
+        - sample_rate_hz: sample rate in hertz
+        """
+
+        super().__init__()
+
+        self._processors = []
+
+        self._chunk_bytes = int(self.CHUNK_S * sample_rate_hz) * channels * bytes_per_sample
+
+        self._cmd = [
+            'arecord',
+            '-q',
+            '-t', 'raw',
+            '-D', input_device,
+            '-c', str(channels),
+            '-f', aiy._drivers._alsa.sample_width_to_string(bytes_per_sample),
+            '-r', str(sample_rate_hz),
+        ]
+        self._arecord = None
+        self._closed = False
+
+    def add_processor(self, processor):
+        """Adds an audio processor.
+
+        An audio processor is an object that has an 'add_data' method with the
+        following signature:
+        class MyProcessor(object):
+          def __init__(self):
+            ...
+
+          def add_data(self, data):
+            # processes the chunk of data here.
+
+        The added processor may be called multiple times with chunks of audio data.
+        """
+        self._processors.append(processor)
+
+    def remove_processor(self, processor):
+        """Removes an added audio processor."""
+
+        if processor in self._processors:
+            self._processors.remove(processor)
+        else:
+            logger.warn("processor was not found in the list")
```

Comment:
if not self._processors.pop(processor, None):
  logger.warn("processor was not found in the list")

avoid scanning the dict twice.